src/cts/src/TreeBuilder.cpp

```diff
@@ -118,42 +156,136 @@ Point<double> TreeBuilder::legalizeOneBuffer(Point<double> bufferLoc,
+      double dist = 0.0;
+      Point<double> noBest(-1e6, -1e6);
+      Point<double> bestLoc = noBest;
+
+      for (const Point<double>& candidate : candidates) {
+        if (!isOccupiedLoc(candidate)) {
+          dist = candidate.computeDist(bufferLoc);
+          if (dist < minDist) {
+            minDist = dist;
+            bestLoc = candidate;
+          }
+        }
       }
-      // fourth, try snapping it to the top
-      delta = y2 - bufferLoc.getY();
-      if (delta < minDist) {
-        newLoc.setX(bufferLoc.getX());
-        newLoc.setY(y2);
+
+      if (bestLoc != noBest) {
+        return bestLoc;
       }
-      return newLoc;
     }
   }
+
   return bufferLoc;
 }
 
+// Check if a particular location is legal by checking
+// 1) if the location is occupied by another cell
+// 2) if the location is sitting on a blockage
+bool TreeBuilder::checkLegalityLoc(const Point<double>& bufferLoc,
+                                   int scalingFactor)
+{
+  // check if location is already occupied
+  if (occupiedLocations_.find(bufferLoc) != occupiedLocations_.end()) {
+    // clang-format off
+    debugPrint(logger_, CTS, "legalizer", 4, "loc {} is already occupied", bufferLoc);
+    // clang-format on
+    return false;
+  }
+
+  double x1, y1, x2, y2;
+  if (findBlockage(bufferLoc, scalingFactor, x1, y1, x2, y2)) {
+    // clang-format off
+    debugPrint(logger_, CTS, "legalizer", 4, "loc {} is in blockage ({:0.3f}"
+	       "{:0.3f}) ({:0.3f} {:0.3f})", bufferLoc, x1, y1, x2, y2);
+    // clang-format on
+
+    return false;
+  }
+
+  return true;
+}
+
+bool TreeBuilder::isOccupiedLoc(const Point<double>& bufferLoc)
+{
+  // clang-format off
+  if (occupiedLocations_.find(bufferLoc) != occupiedLocations_.end()) {
+    debugPrint(logger_, CTS, "legalizer", 4, "loc {} is already occupied", bufferLoc);
+    return true;
+  }
+  debugPrint(logger_, CTS, "legalizer", 4, "loc {} is not occupied", bufferLoc);
+  return false;
+  // clang-format on
+}
+
+void TreeBuilder::commitLoc(const Point<double>& bufferLoc)
+{
+  // clang-format off
+  occupiedLocations_.insert(bufferLoc);
+  debugPrint(logger_, CTS, "legalizer", 4, "loc {} has been committed, size={}",
+             bufferLoc, occupiedLocations_.size());
+  // clang-format on
+}
+
+void TreeBuilder::uncommitLoc(const Point<double>& bufferLoc)
+{
+  // clang-format off
+  occupiedLocations_.erase(bufferLoc);
+  debugPrint(logger_, CTS, "legalizer", 4, "loc {} has been uncommitted, size={}",
+             bufferLoc, occupiedLocations_.size());
+  // clang-format on
+}
+
+void TreeBuilder::commitMoveLoc(const Point<double>& oldLoc,
+                                const Point<double>& newLoc)
+{
+  // clang-format off
+  occupiedLocations_.erase(oldLoc);
+  occupiedLocations_.insert(newLoc);
+  debugPrint(logger_, CTS, "legalizer", 4, "move:{} -> {} has been committed, size={}",
+             oldLoc, newLoc, occupiedLocations_.size());
+  // clang-format on
```

Comment:
Same clang-format comment from before, here and above.

Reply:
Same disagreement as ablove.